### PR TITLE
Refactor definition of Module to use pathtype

### PR DESCRIPTION
**Pull request by zhujinxuan**
Work in progress, Shall close https://github.com/github/semantic/issues/522

### CODE REVIEW

File: src/Data/Abstract/Path.hs [hunk 1]
```diff
@@ -31,9 +31,9 @@ joinPaths a b = let bs = splitPath (normalise b)
 --     joinPaths "a/b" "./c" == "a/b/c"
 --
 -- Walking beyond the beginning of a just stops when you get to the root of a.
--- JZ:TODO: Refactor later
-pjoin :: FileDir fd => Path.AbsRelDir -> Path.Rel fd -> Path.AbsRel fd
-pjoin x y= Path.path $ joinPaths (Path.toString x) (Path.toString y)
+-- TODO: Rewrite it with pathtype
```
Comment (zhujinxuan) [inline src/Data/Abstract/Path.hs:34]: I will rewrite the joinPaths with tests in another PR.

File: src/Control/Abstract/Modules.hs [hunk 2]
```diff
@@ -157,7 +157,6 @@ data ResolutionError resume where
                 -> Language -- Language.
                 -> ResolutionError ModulePath
 
-  -- JZ:Shall it be FilePath? as this first value can be github.com/... for GoLang
   GoImportError :: FilePath -> ResolutionError [ModulePath]
```
Comment (zhujinxuan) [inline src/Control/Abstract/Modules.hs]: Perhaps it is better to write String here.  GoImportError can be followed with an uri like `https://github.com/....`, which it is not a file path.

File: src/Language/Python/Syntax.hs [hunk 3]
```diff
@@ -96,7 +96,6 @@ resolvePythonModules q = do
 
     search rootDir x = do
       trace ("searching for " <> show x <> " in " <> show rootDir)
-    -- JZ:TODO: Refactor later
       let path = normalise (rootDir </> normalise x)
```
Comment (zhujinxuan) [inline src/Language/Python/Syntax.hs:99]: This is related to https://github.com/github/semantic/issues/537.  I will refactor it in another PR.

File: src/Language/TypeScript/Resolution.hs [hunk 4]
```diff
@@ -110,7 +110,6 @@ resolveModule :: ( Has (Modules address value) sig m
               -> [String] -- ^ File extensions to look for
               -> Evaluator term address value m (Either [M.ModulePath] M.ModulePath)
 resolveModule path' exts = do
-  -- JZ:TODO: Refactor later
   let path = makeRelative "." $ Path.toString path'
```
Comment (zhujinxuan) [inline src/Language/TypeScript/Resolution.hs:113]: This is also related to https://github.com/github/semantic/issues/537.  I will refactor it in another PR.

File: test/Data/Abstract/Path/Spec.hs [hunk 5]
```diff
@@ -7,14 +7,14 @@ spec :: Spec
 spec = parallel $
   describe "joinPaths" $ do
     it "joins empty paths" $
-      joinPaths "" "" `shouldBe` "."
+      joinPathf "" "" `shouldBe` "."
```
Comment (zhujinxuan) [inline test/Data/Abstract/Path/Spec.hs:10]: I will implement the test of joinPaths in another PR along with joinPaths rewrite.

File: src/Control/Abstract/Modules.hs (hunk 2 above)
Comment (patrickt) [reply src/Control/Abstract/Modules.hs]: @zhujinxuan Yup, `String`, not `FilePath` or a typed path is correct here. Can you change it to `String` and add a comment explaining why this is so?

File: src/Language/TypeScript/Resolution.hs [hunk 6]
```diff
@@ -83,39 +83,40 @@ resolveNonRelativePath :: ( Has (Modules address value) sig m
                           , Has (Resumable (BaseError ResolutionError)) sig m
                           , Has Trace sig m
                           )
-                       => FilePath
+                       => Path.RelDir
```
Comment (patrickt) [inline src/Language/TypeScript/Resolution.hs:86]: The name of this function is a little weird given that it says “NonRelative” but takes a RelDir. Can you add a TODO about renaming it?

File: src/Semantic/Graph.hs [hunk 7]
```diff
@@ -383,7 +384,7 @@ resumingResolutionError :: ( Has Trace sig m
 resumingResolutionError = runResolutionErrorWith $ \ baseError -> do
   traceError "ResolutionError" baseError
   case baseErrorException baseError of
-    NotFoundError nameToResolve _ _ -> pure $ Path.absRel nameToResolve
+    NotFoundError nameToResolve _ _ -> pure $ fromMaybe (Path.toAbsRel Path.emptyFile) $ Path.fileFromFileDir nameToResolve
```
Comment (patrickt) [inline src/Semantic/Graph.hs:387]: The empty path isn’t strictly correct here, but I’m not sure of the ramifications of that. For now, can you call `trace` or `traceError` here with information about this name that isn’t structurally valid?

File: src/Data/Abstract/Path.hs [hunk 8]
```diff
@@ -2,26 +2,40 @@ module Data.Abstract.Path
...
 -- | Join two paths a and b. Handles walking up relative directories in b. e.g.
 --
 --     joinPaths "a/b" "../c" == "a/c"
 --     joinPaths "a/b" "./c" == "a/b/c"
 --
 -- Walking beyond the beginning of a just stops when you get to the root of a.
-joinPaths :: FilePath -> FilePath -> FilePath
-joinPaths a b = let bs = splitPath (normalise b)
+joinPathf :: FilePath -> FilePath -> FilePath
+joinPathf a b = let bs = splitPath (normalise b)
```
Comment (patrickt) [inline src/Data/Abstract/Path.hs:21]: This name is a little opaque; I would call it `joinUntypedPaths` or something, and insert a comment that says to prefer `joinPaths` if humanly possible.

File: src/Language/TypeScript/Resolution.hs (hunk 6 above)
Comment (zhujinxuan) [reply src/Language/TypeScript/Resolution.hs:90]: Fixed!

File: src/Semantic/Graph.hs (hunk 7 above)
Comment (zhujinxuan) [reply src/Semantic/Graph.hs:387]: I think the best way to fix it is to change the definition of ResolutionError as 
```
data ResolutionError resume where
  NotFoundError :: Path.AbsRelFileDir   -- The path that was not found.
                -> [Path.AbsRelFile] -- List of paths searched that shows where semantic looked for this module.
                -> Language -- Language.
                -> ResolutionError Path.AbsRelDir -- which is ModulePath currently.
``` 

If you agree with this fix, I will make a TODO here and fix it in next PR or this PR.

Comment (zhujinxuan) [reply src/Semantic/Graph.hs:387]: I think, respectively, `GoImportError` shall be typed as 

```
GoImportError :: String -> ResolutionError [String]
```

Comment (zhujinxuan) [reply src/Semantic/Graph.hs:387]: Hi, this is about something in pathtype.  The only possibility of "Path.fileFromFileDir nameToResolve" is that the nameToResolve is empty, like `Path.currentDir`.

I have done some tests in ghci, which shows some subtle differences 
```
> Path.fileFromFileDir $ Path.toFileDir $ Path.currentDir 
Nothing
> Path.fileFromFileDir $ Path.toFileDir $ Path.relDir "./"
Just (currentDir </> relPath ".")
> Path.fileFromFileDir $ Path.toFileDir $ Path.relDir "."
Just (currentDir </> relPath ".")
```

File: src/Data/Abstract/Path.hs (hunk 8 above)
Comment (zhujinxuan) [reply src/Data/Abstract/Path.hs]: Fixed!

File: src/Semantic/Graph.hs (hunk 7 above)
Comment (zhujinxuan) [reply src/Semantic/Graph.hs:387]: I think the problem is that we want to resume with a ModulePath anyway.  Perhaps, in the future, we can think about define module path with status.
``
  data Status = Resolved | NotFound | Undecided
  data ModulePath (x :: Status ) where
      Resolved :: (ResolvedOrUndecided x) Path.AbsRelFile -> ModulePath x
      NotFound :: (NotFoundOrUndecided x) String -> ModulePath x
```

Then we can resume with `ModulePath 'Undecided`, and process blob with `ModulePath 'Resolved`.  And then make a simple adapter between these two.

I am not sure whether this design may indicate a fused-effect library.

Comment (zhujinxuan) [reply src/Semantic/Graph.hs]: A traceError temporary fix is added by c2542a1

**Comment by zhujinxuan** [comment]
Hi, @patrickt.  When you merge this PR, can you use `squash and merge`?   i would like to put all modification in one commit, as it appears nice in commit history.

**Comment by patrickt** [comment]
@zhujinxuan Our position on the semantic team is that all history is worth capturing, even churn on review branches: the task of handling that complexity when exploring git history is the responsibility of our tools, not our workflows. (Not a perfect workflow, I’ll admit, but consistency here is the most important thing.) [Here](https://github.com/github/semantic/pull/380) is an example of a PR that had a ton of post-hoc fixes. 

Thank you again! As always, we truly appreciate your contribution.

**Comment by patrickt** [comment]
@zhujinxuan Our position on the semantic team is that all history is worth capturing, even churn on review branches: the task of handling that complexity when exploring git history is the responsibility of our tools, not our workflows. (Not a perfect workflow, I’ll admit, but consistency here is the most important thing.) [Here](https://github.com/github/semantic/pull/380) is an example of a PR that had a ton of post-hoc fixes.

**Pull request merged**